src/nma/nma.rs

```diff
@@ -0,0 +1,412 @@
+pub fn write_u32(mut file: &File, data: u32) -> std::io::Result<()> {
+    const PADDED_LENGTH: usize = 4;
+
+    // Create a 4-length buffer
+    let mut padded_buffer: [u8; PADDED_LENGTH] = [0; PADDED_LENGTH];
+
+    // Convert data into bytes
+    let bytes: [u8; PADDED_LENGTH] = data.to_be_bytes();
+
+    // Copy bytes into buffer at offset s.t. value is left-padded with 0s
+    let copy_index = PADDED_LENGTH - bytes.len();
+    padded_buffer[copy_index..].copy_from_slice(&bytes);
+
+    // Write to file
+    file.write_all(&padded_buffer)
+}
+
+/// Represents the state of the assembler at any given point
+#[derive(Debug, PartialEq)]
+enum AssemblerState {
+    /// State before any processing has occurred
+    Initial,
+    /// The assembler is in the process of scanning in new tokens
+    Scanning,
+    /// The assembler has encountered an R-type instruction and
+    /// is collecting its arguments before assembling
+    CollectingRArguments,
+    /// The assembler has encountered an I-type instruction and
+    /// is collecting its arguments before assembling
+    CollectingIArguments,
+}
+
+/// Converts a numbered mnemonic ($t0, $s8, etc) or literal (55, 67, etc) to its integer representation
+fn reg_number(mnemonic: &str) -> Result<u32, &'static str> {
+    if mnemonic.len() != 3 {
+        return Err("Mnemonic out of bounds");
+    }
+    match mnemonic.chars().nth(2) {
+        Some(c) => match c.to_digit(10) {
+            Some(digit) => Ok(digit),
+            _ => Err("Invalid register index"),
+        },
+        _ => Err("Malformed mnemonic"),
+    }
+}
+
+/// Given a register or number, assemble it into its integer representation
+fn assemble_reg(mnemonic: &str) -> Result<u32, &'static str> {
+    if !mnemonic.starts_with("$") {
+        return match mnemonic.parse::<u32>() {
+            Ok(v) => Ok(v),
+            Err(_) => Err("Failed to parse shamt"),
+        };
+    }
+
+    // match on everything after $
+    match &mnemonic[1..] {
+        "zero" => Ok(0),
+        "at" => Ok(1),
+        "gp" => Ok(28),
+        "sp" => Ok(29),
+        "fp" => Ok(30),
+        "ra" => Ok(31),
+        _ => {
+            let n = reg_number(mnemonic)?;
+            let reg = match mnemonic.chars().nth(1) {
+                Some('v') => 2 + n,
+                Some('a') => 4 + n,
+                Some('t') => {
+                    if n <= 7 {
+                        8 + n
+                    } else {
+                        // t8, t9 = 24, 25
+                        // 24 - 8 + n
+                        16 + n
+                    }
+                }
+                Some('s') => 16 + n,
+                _ => 99,
+            };
+            if reg <= 31 {
+                Ok(reg)
+            } else {
+                Err("Register out of bounds")
+            }
+        }
+    }
+}
+
+/// Assembles an R-type instruction
+fn assemble_r(r_struct: &mut R, r_args: Vec<&str>) -> Result<u32, &'static str> {
+    let mut rs: u32 = 0;
```

Comment:
You wrote a U5 and U6 type with a complete implementation, and then used u32s for the registers?